File: day2.swift
#! /usr/bin/swift

func run(_ code: inout [Int], _ noun: Int, _ verb: Int) -> Int {
  var ip = 0
  code[1] = noun
  code[2] = verb

  while code[ip] != 99 {
    if code[ip] == 1 {
      code[code[ip + 3]] = code[code[ip + 1]] + code[code[ip + 2]]
    } else {
      code[code[ip + 3]] = code[code[ip + 1]] * code[code[ip + 2]]
    }
    ip += 4
  }
  return code[0]
}

var code = readLine()!.split(separator: ",").map { Int($0)! }

var memory = code
print(run(&memory, 12, 2))

outer: for noun in stride(from: 0, through: 100, by: 1) {
  for verb in stride(from: 0, through: 100, by: 1) {
    memory = code
    if run(&memory, noun, verb) == 19690720 {
      print("\(noun * 100 + verb)")
      break outer
    }
  }
}

File: day2.rb
#! /usr/bin/env ruby

# frozen_string_literal: true

code = $stdin.readline.split(',').map(&:to_i)
code[1] = 12
code[2] = 2

ip = 0
while code[ip] != 99
  location1 = code[ip + 1]
  location2 = code[ip + 2]
  store = code[ip + 3]

  if code[ip] == 1
    code[store] = code[location1] + code[location2]
  elsif code[ip] == 2
    code[store] = code[location1] * code[location2]
  else
    puts "Bad opcode #{code[ip]} at #{ip}"
    exit(1)
  end

  ip += 4
end

puts code[0]
